### PR TITLE
Option to log Maestro request-times.

### PR DESCRIPTION
**Pull request by ajuggler**
Thank you for contributing to Atlas! We appreciate your effort and dedication to improving this project. To ensure that your contribution is in line with the project's guidelines and can be reviewed efficiently, please fill out the template below.

Remember to follow our [Contributing Guide](CONTRIBUTING.md) before submitting your pull request.

## Summary

Please provide a brief, concise summary of the changes in your pull request. Explain the problem you are trying to solve and the solution you have implemented.

## Type of Change

Please mark the relevant option(s) for your pull request:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code refactoring (improving code quality without changing its behavior)
- [ ] Documentation update (adding or updating documentation related to the project)

## Checklist

Please ensure that your pull request meets the following criteria:

- [ ] I have read the [Contributing Guide](CONTRIBUTING.md)
- [ ] My code follows the project's coding style and best practices
- [ ] My code is appropriately commented and includes relevant documentation
- [ ] I have added tests to cover my changes
- [ ] All new and existing tests pass
- [ ] I have updated the documentation, if necessary

## Testing

Please describe the tests you have added or modified, and provide any additional context or instructions needed to run the tests.

- [ ] Test A
- [ ] Test B

## Additional Information

If you have any additional information or context to provide, such as screenshots, relevant issues, or other details, please include them here.

### CODE REVIEW

File: src/GeniusYield/GYConfig.hs [hunk 1]
```diff
@@ -119,9 +119,10 @@ In JSON format, this essentially corresponds to:
 = { coreProvider: GYCoreProviderInfo, networkId: NetworkId, logging: [GYLogScribeConfig], utxoCacheEnable: boolean }
 -}
 data GYCoreConfig = GYCoreConfig
-  { cfgCoreProvider :: !GYCoreProviderInfo
-  , cfgNetworkId    :: !GYNetworkId
-  , cfgLogging      :: ![GYLogScribeConfig]
+  { cfgCoreProvider  :: !GYCoreProviderInfo
+  , cfgNetworkId     :: !GYNetworkId
+  , cfgLogging       :: ![GYLogScribeConfig]
+  , cfgMaestroTiming :: !Bool
```
Comment (sourabhxyz) [inline src/GeniusYield/GYConfig.hs:125]: This has nothing to do with Maestro and is to be applicable for all providers.

Comment (ajuggler) [reply src/GeniusYield/GYConfig.hs]: OK, even simpler.

Note that I have kept the commented lines 209-215.  Without these and intermediate variables `(gyQueryUTxO', gySlotActions')`, code could be simplified.

File: src/GeniusYield/GYConfig.hs [hunk 2]
```diff
@@ -203,15 +205,25 @@ withCfgProviders
...
+              let gySlotActions = gySlotActions' { gyWaitForNextBlock' = purgeCache >> gyWaitForNextBlock' gySlotActions'}
+              pure (gyQueryUTxO, gySlotActions, f)
+          else pure (gyQueryUTxO', gySlotActions', f) -}
+          if cfgLogTiming then do
+              let fT = \provider -> do
+                    start <- getCurrentTime
+                    a <- f provider
+                    end   <- getCurrentTime
+                    let duration = diffUTCTime end start
+                    logRun gyLog' mempty GYDebug $ printf "Maestro took: " ++ show duration
```
Comment (sourabhxyz) [inline src/GeniusYield/GYConfig.hs:222]: Shouldn't say "Maestro" here. Preferably, something on lines, "Request took: "

File: src/GeniusYield/GYConfig.hs [hunk 3]
```diff
@@ -203,15 +205,25 @@ withCfgProviders
...
+          then do
+              (gyQueryUTxO, purgeCache) <- CachedQuery.makeCachedQueryUTxO gyQueryUTxO' gyLog'
+              -- waiting for the next block will purge the utxo cache.
+              let gySlotActions = gySlotActions' { gyWaitForNextBlock' = purgeCache >> gyWaitForNextBlock' gySlotActions'}
+              pure (gyQueryUTxO, gySlotActions, f)
+          else pure (gyQueryUTxO', gySlotActions', f) -}
+          if cfgLogTiming then do
+              let fT = \provider -> do
+                    start <- getCurrentTime
+                    a <- f provider
```
Comment (sourabhxyz) [inline src/GeniusYield/GYConfig.hs:219]: Hm.. Need to think if this is appropriate. You are wrapping around general call to provider and not to call of individual query calls.

Comment (ajuggler) [reply src/GeniusYield/GYConfig.hs]: Time logs are now added to individual query calls.

File: src/GeniusYield/GYConfig.hs (hunk 2 above)
Comment (ajuggler) [reply src/GeniusYield/GYConfig.hs]: Done.

File: src/GeniusYield/GYConfig.hs [hunk 4]
```diff
@@ -208,17 +210,98 @@ withCfgProviders
...
+        , gyQueryUtxoRefsAtAddress'             = wrap "gyQueryUtxoRefsAtAddress" . gyQueryUtxoRefsAtAddress providers
+        , gyQueryUtxosAtAddress'                = \addr mac -> wrap "gyQueryUtxosAtAddress'" $ gyQueryUtxosAtAddress providers addr mac
+        , gyQueryUtxosAtAddressWithDatums'      = case gyQueryUtxosAtAddressWithDatums' gyQueryUTxO of
+            Nothing -> Nothing
+            Just q  -> Just $ \addr mac -> wrap "gyQueryUtxosAtAddressWithDatums'" $ q addr mac
+        , gyQueryUtxosAtAddresses'              = wrap "gyQueryUtxosAtAddresses" . gyQueryUtxosAtAddresses providers
+        , gyQueryUtxosAtAddressesWithDatums'    = case gyQueryUtxosAtAddressesWithDatums' gyQueryUTxO of
+            Nothing -> Nothing
+            Just q  -> Just $ wrap "gyQueryUtxosAtAddressesWithDatums" . q
+        , gyQueryUtxosAtPaymentCredential'      = \cred -> wrap "gyQueryUtxosAtTxOutRefs" . gyQueryUtxosAtPaymentCredential providers cred
```
Comment (sourabhxyz) [inline src/GeniusYield/GYConfig.hs:289]: Seems like a typo here.

**Review by sourabhxyz — CHANGES_REQUESTED** [review]

**Review by sourabhxyz — APPROVED** [review]

**Pull request merged**